### ISSUE TITLE
Bazel test failure: Lots of h2_oauth2_tests fail under msan

### ISSUE THREAD

**Issue by adelez** [grpc/grpc#15611]
One example: https://source.cloud.google.com/results/invocations/b4505ae7-d0d0-4a97-8f1a-f43c8cefb193/targets/%2F%2Ftest%2Fcore%2Fend2end:h2_oauth2_test@bad_ping@poller%3Depoll1/log

Here you can see a list of failing h2_oauth2_tests: https://source.cloud.google.com/results/invocations/b4505ae7-d0d0-4a97-8f1a-f43c8cefb193/targets

**Comment by adelez** [comment]
Same with h2_ssl_proxy_test, h2_ssl_session_reuse_test, h2_ssl_test.

**Comment by jiangtaoli2016** [comment]
@adelez when do you start noticing such failure?

**Comment by jiangtaoli2016** [comment]
According to @davidben, we need to turn **off** assembly when building msan.

**Comment by davidben** [comment]
MSan works by instrumenting all memory accesses, notably writes, so it knows when memory is initialized. If you call into uninstrumented code to initialize something, MSan doesn't know it ever was initialized. That's why MSan requires that *all* dependencies be built with MSan.

BoringSSL includes a lot of assembly functions. Those are, of course, not compiled and thus MSan cannot instrument them. So whenever you run tests under MSan in general, you must turn off assembly in all your projects.

https://clang.llvm.org/docs/MemorySanitizer.html#handling-external-code

**Comment by adelez** [comment]
Hi Jiangtao, I just got MSAN to work with bazel/Foundry today and found these failures. So it may have always been there. With bazel and Foundry, we use the latest Clang, which can catch new failures that older version didn't.

**Comment by jiangtaoli2016** [comment]
@nicolasnoble Can you help to set the correct msan bazel build rule?

**Comment by nicolasnoble** [comment]
@davidben right - but what's the Bazel target for this ? Or in other words, when I look at https://github.com/google/boringssl/blob/chromium-stable-with-bazel/BUILD I don't see any good way to do this from our side of the build.

**Comment by davidben** [comment]
I think you all might be the first to try to use that file with MSan. :-)

I'm not familiar with bazel, so I don't know how its configurations and build toggles work. I assume it should either condition on a standard is_msan toggle, if it exists, or otherwise export some no_asm toggle for the user to flip based on their own is_msan toggle. Or something like that? Feel free to send over some appropriate patch.

**Comment by nicolasnoble** [comment]
Right, so what I'm hearing is that the BoringSSL's Bazel files haven't been tuned around testing with msan, which really is the rootcause here.

I'm not a Bazel expert either, despite people thinking that I am :P We'll have to discuss with the Bazel team to figure out how we can solve this exactly.
